### PR TITLE
Fixed #34266 -- Added ClosestPoint GIS database functions.

### PR DESCRIPTION
**Pull request by niccolomineo**
I'd like to suggest adding this function to the bunch to encourage people to prefer its use in a number of situations over [cumbersome solutions](https://niccolomineo.com/articles/marker-center-complex-line-geodjango/).

### CODE REVIEW

**Comment by pauloxnet** [comment]
Please add references to this new feature in the documentation as well especially in:

https://github.com/django/django/blob/main/docs/ref/contrib/gis/functions.txt

and

https://github.com/django/django/blob/main/docs/ref/contrib/gis/db-api.txt

See a similar function for reference.

**Review by emarhin — APPROVED** [review]
👍

File: django/contrib/gis/db/backends/spatialite/operations.py [hunk 1]
```diff
@@ -80,7 +80,13 @@ class SpatiaLiteOperations(BaseSpatialOperations, DatabaseOperations):
 
     @cached_property
     def unsupported_functions(self):
-        unsupported = {"BoundingCircle", "GeometryDistance", "IsEmpty", "MemSize"}
+        unsupported = {
+            "BoundingCircle",
+            "ClosestPoint",
```
Comment (pauloxnet) [inline django/contrib/gis/db/backends/spatialite/operations.py:85]: Actually it seems that Spatialite support it
https://www.gaia-gis.it/gaia-sins/spatialite-sql-5.0.1.html#p14b

**Review by pauloxnet — CHANGES_REQUESTED** [review]

File: docs/ref/contrib/gis/functions.txt [hunk 2]
```diff
@@ -237,6 +237,15 @@ The ``num_seg`` parameter is used only on PostGIS.
 Accepts a single geographic field or expression and returns the ``centroid``
 value of the geometry.
 
+``ClosestPoint``
+================
+
+.. class:: ClosestPoint(expr1, expr2, **extra)
+
+*Availability*: `PostGIS <https://postgis.net/docs/ST_ClosestPoint.html>`__, SpatiaLite
```
Comment (pauloxnet) [inline docs/ref/contrib/gis/functions.txt:245]:
```suggestion
*Availability*: `PostGIS <https://postgis.net/docs/ST_ClosestPoint.html>`__, `SpatiaLite <https://www.gaia-gis.it/gaia-sins/spatialite-sql-5.0.1.html#p14b>`__
```

**Comment by pauloxnet** [comment]
buildbot, test on oracle.

**Comment by niccolomineo** [comment]
buildbot, test on oracle.

**Comment by niccolomineo** [comment]
buildbot, test on oracle.

**Comment by niccolomineo** [comment]
buildbot, test on oracle.

**Comment by niccolomineo** [comment]
buildbot, test on oracle.

**Comment by niccolomineo** [comment]
buildbot, test on oracle.

**Comment by niccolomineo** [comment]
buildbot, test on oracle.

**Review by carltongibson — COMMENTED** [review]
Hi @niccolomineo, thanks for this! 

Can you add a release note to the `django.contrib.gis` section of `docs/releases/5.0.txt`. (See similar in 4.2.txt for a model.)

**Comment by niccolomineo** [comment]
> Hi @niccolomineo, thanks for this!
> 
> Can you add a release note to the `django.contrib.gis` section of `docs/releases/5.0.txt`. (See similar in 4.2.txt for a model.)

Just added!

**Review by pauloxnet — APPROVED** [review]
These changes look to me ready to squash commits and merge.

File: docs/ref/contrib/gis/functions.txt [hunk 3]
```diff
@@ -237,6 +237,15 @@ The ``num_seg`` parameter is used only on PostGIS.
 Accepts a single geographic field or expression and returns the ``centroid``
 value of the geometry.
 
+``ClosestPoint``
+================
+
```
Comment (felixxm) [inline docs/ref/contrib/gis/functions.txt:242]: `versionadded` annotation is missing.

File: docs/ref/contrib/gis/functions.txt [hunk 4]
```diff
@@ -237,6 +237,15 @@ The ``num_seg`` parameter is used only on PostGIS.
...
+.. class:: ClosestPoint(expr1, expr2, **extra)
+
+*Availability*: `PostGIS <https://postgis.net/docs/ST_ClosestPoint.html>`__, `SpatiaLite <https://www.gaia-gis.it/gaia-sins/spatialite-sql-5.0.1.html#p14b>`__
+
+Accepts two geographic fields or expressions and returns the geometry A's 2-dimensional closest point to geometry B.
```
Comment (felixxm) [inline docs/ref/contrib/gis/functions.txt:247]: Wrap docs at 79 chars.

File: docs/releases/5.0.txt [hunk 5]
```diff
@@ -62,7 +62,9 @@ Minor features
 :mod:`django.contrib.gis`
 ~~~~~~~~~~~~~~~~~~~~~~~~~
 
-* ...
+* The new :class:`ClosestPoint() <django.contrib.gis.db.models.functions.ClosestPoint>`
+  function allows returning a geometry's 2-dimensional point that is closest to another
+  geometry.
```
Comment (felixxm) [inline docs/releases/5.0.txt:67]: Wrap docs at 79 chars.

File: docs/ref/contrib/gis/functions.txt [hunk 6]
```diff
@@ -237,6 +237,15 @@ The ``num_seg`` parameter is used only on PostGIS.
 Accepts a single geographic field or expression and returns the ``centroid``
 value of the geometry.
 
+``ClosestPoint``
+================
+
+.. class:: ClosestPoint(expr1, expr2, **extra)
+
+*Availability*: `PostGIS <https://postgis.net/docs/ST_ClosestPoint.html>`__, `SpatiaLite <https://www.gaia-gis.it/gaia-sins/spatialite-sql-5.0.1.html#p14b>`__
```
Comment (felixxm) [inline docs/ref/contrib/gis/functions.txt:245]: Is it available in SpatiaLite 4.3.0?

Comment (niccolomineo) [reply docs/ref/contrib/gis/functions.txt:245]: It is, I am replacing the reference URL.

**Comment by niccolomineo** [comment]
@felixxm let me know if anything else is missing.

**Comment by felixxm** [comment]
@niccolomineo Thanks :+1: Welcome aboard :boat: 

I pushed small edits.

**Pull request merged**